CedarJava/src/main/java/com/cedarpolicy/model/PartialAuthorizationResponse.java

```diff
@@ -0,0 +1,119 @@
+package com.cedarpolicy.model;
+
+import com.cedarpolicy.model.AuthorizationResponse.Decision;
+import com.cedarpolicy.model.AuthorizationResponse.Diagnostics;
+import com.cedarpolicy.model.slice.Policy;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public abstract class PartialAuthorizationResponse {
+    private final Diagnostics diagnostics;
+
+    @JsonCreator
+    public static PartialAuthorizationResponse createPartialAuthorizationResponse(
+            @JsonProperty("response") PartialAuthorizationResponse nested,
+            @JsonProperty("decision") Decision decision,
+            @JsonProperty("residuals") Map<String, JsonNode> residuals,
+            @JsonProperty("diagnostics") Diagnostics diagnostics) {
+        if (nested != null) {
```

Comment:
It's a little trickery of the JSON deserialization framework. This method is registered as the factory method for deserializing PartialAuthorizationResponse objects. When we get the response from the JNI side, the JSON deserializer finds this factory method, based on its parameters it finds the "response" attribute, but no "decision", "residuals", or "diagnostics" ones. For the "response" attribute, it sees that the associated JSON object should be deserialized as a PartialAuthorizationResponse so it uses this factory method again. This time, it finds no "response" attribute, but if finds a "diagnostics" attribute and either a "decision" or "residuals" one. It deserializes all these parameters using the respective object constructors and calls the factory method with the constructed objects. The logic in the factory method  then instantiate a ConcretePartialAuthorizationResponse or ResidualPartialAuthorizationResponse depending on the based on the non-null parameters and return it. Finally, it can now invoke the factory method for the root JSON object with nested set to the constructed object and the method will return it verbatim.

Reply:
There might be a more elegant way to basically ignore the root node of the JSON response but it avoids needing a bunch of private inner classes just to support deserialization.